### PR TITLE
Substitute `ChallengeGenerator` by the generic sponge

### PR DESCRIPTION
**Pull request by autquis**
## Description

In this PR, we are removing `ChallengeGenerator` and using only the sponge on which the PCS is generic. As a consequence, we are dropping the two strategies of generating challenges. However, as far as I noticed, the `Univariate` strategy was only used in a few tests and bench templates. So, I am not sure how big of an issue it is.

Closes #138 
Notice that this PR is on top of #137 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: poly-commit/src/lib.rs [hunk 1]
```diff
@@ -665,12 +656,13 @@ pub mod tests {
...
-            ChallengeGenerator::new_multivariate(sponge()),
-            ChallengeGenerator::new_univariate(&mut sponge()),
+        let sponges = vec![
+            sponge(),
+            // ChallengeGenerator::new_multivariate(sponge()),
+            // ChallengeGenerator::new_univariate(&mut sponge()),
         ];
 
-        for challenge_gen in challenge_generators {
+        for sponge in sponges {
```
Comment (autquis) [inline poly-commit/src/lib.rs:665]: This loop is extra now, but I will wait for your confirmation.

File: poly-commit/src/lib.rs [hunk 2]
```diff
@@ -665,12 +654,9 @@ pub mod tests {
...
         S: CryptographicSponge,
     {
-        let challenge_generators = vec![
-            ChallengeGenerator::new_multivariate(sponge()),
-            ChallengeGenerator::new_univariate(&mut sponge()),
-        ];
+        let sponge = sponge();
 
-        for challenge_gen in challenge_generators {
+        for __ in 0..1 {
```
Comment (autquis) [inline poly-commit/src/lib.rs:659]: By deleting this for loop, the diff gets unnecessarily complicated. As soon as the PR gets a review, I will delete this for loop, and then please merge it.

**Comment by Pratyush** [comment]
Thanks for this PR! It looks pretty good overall. Since you're looking into this code, could you double check that the way we're deriving these challenges is not vulnerable to  https://blog.openzeppelin.com/the-last-challenge-attack ?

Essentially, when we're doing batch verification, we need to ensure that we absorb all the proofs in the batch before we squeeze the `opening_challenge`.

**Comment by autquis** [comment]
Of course!

- Currently, the sponge does not absorb anything. Even before this PR, the `ChallengeGenerator` was not absorbing anything. Is this a problem in Marlin and other schemes in the crate? (I am unfamiliar with their security analysis)
- For Ligero, Brakedown, and Hyrax, I am not sure how we can absorb all the transcripts in `batch_check`. As the current implementation of `batch_check` calls `check` in a for loop, and the `check` method should absorb the transcript anyhow, I do not see an immediate solution to avoid double-absorbing. 🤔 

Thanks!

**Comment by mmagician** [comment]
True, there was only ever squeezes from a sponge...

Regarding the second point for IOP-based schemes, we anyway planned to improve the `batch_check` further down the line.

**Comment by autquis** [comment]
Thanks for the answer. Regarding `batch_check`: One rough idea is to add another method (e.g., `check_with_challange`), and this method does not get the sponge as an argument but the challenges are given. And `batch_check` absorbs everything and squeezes all necessary challenges for each `check_with_challange`.
I think this can go in another PR, though.

**Comment by autquis** [comment]
Thanks for the answer! Regarding `batch_verify`: Maybe the following rough idea helps. Instead of having `check`, we use two methods, `check_transcript` and `check_proof`. `check_transcript` has the proof and the challenges as arguments (i.e., no absorbing or squeezing). `check_proof` has the proof and the sponge as arguments, and needs to absorb or squeeze, then call `check_transcript`. `batch_check` does the same job as `check_proof` with some potential optimization for the case of many proofs.
I think it requires another discussion/issue/PR, though.

**Comment by Pratyush** [comment]
Ok seems like the blog post and what this PR aim to achieve are two different things; this PR talks about how to generate challenges for batch proving, and the blog post talks about batching pairings *within* a batch proof verification.

That is, the blog post talks about replacing [this](https://github.com/arkworks-rs/poly-commit/blob/c724fa666e935bbba8db5a1421603bab542e15ab/poly-commit/src/kzg10/mod.rs#L390) line with a squeeze from a sponge instead of a `rng`-derived challenge.

We should change that in a separate PR (it's not a soundness bug atm.)

**Comment by Pratyush** [comment]
Having resolved that confusion (sorry about that!), let's return our focus to the topic of this PR.

I think the reason why the current code is "sound" is that it assumes the sponge inside the challenge generator already absorbed the requisite commitments. For example, see the `marlin` code here: https://github.com/arkworks-rs/marlin/blob/026b73c20638f4f86cbae0946045934c865d5a30/src/lib.rs#L289. The absorption happens before we generate `opening_challenge`.

 We can think about whether we want to change that or not; my inclination is that the answer is yes.

**Comment by mmagician** [comment]
That makes sense @Pratyush. In any case we can tackle this separately as this PR doesn't alter the old behavior right?

**Review by Pratyush — APPROVED** [review]
Yup, looks good. Can we file a follow-up issue?

File: poly-commit/src/lib.rs (hunk 2 above)
Comment (Pratyush) [reply poly-commit/src/lib.rs:659]: PR is approved!

**Pull request merged**